### PR TITLE
Update to use `KeyPath` scoping

### PR DESCRIPTION
**Pull request by rhysm94**
This PR brings TCACoordinators more in-line with TCA in a number of ways.

Primarily, it allows TCACoordinators to use [cached store scoping using KeyPaths](https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/migratingto1.5/#Store-scoping-with-key-paths), as introduced in TCA 1.5, removing usage of deprecated APIs.

Some syntax has been necessarily tweaked to make this happen. First, `TCARouter`'s initialiser has been updated to take a `Store<[Route<Screen>], RouterAction<ID, Screen, ScreenAction>>`. Instead of relying on the `IdentifiedRouterState` and `IndexedRouterState` protocols to generate key paths to the state, instead, users will directly scope their store to the array (or `IdentifiedArray`) of routed screens.

The `IdentifiedRouterAction` and `IndexedRouterAction` protocols have been removed, instead replaced with some type aliases for the new `RouterAction<ID, Screen, ScreenAction>` type. This allows the library to generate the required `CaseKeyPath`s for scoping. This does mean that anywhere users previously had:

```swift
enum Action: IdentifiedRouterAction {
  case routeAction(id: Screen.State.ID, action: Screen.Action)
  case updateRoutes(_ routes: [Route<Screen.State>])
}
```

they need to move to:

```swift
enum Action {
  case router(IdentifiedRouterAction<Screen.State, Screen.Action>
}
```

although I've added `IdentifiedRouterActionOf<Reducer>` and `IndexedRouterActionOf<Reducer>` type aliases as well, so this is valid too:

```swift
enum Action {
  case router(IdentifiedRouterActionOf<Screen>)
}
```

I tried to figure out a way to make this all backwards compatible, but it's basically impossible to generate a `CaseKeyPath` from a protocol function, and if I've got to break compatibility to add `CaseKeyPath` requirements instead of/as well as the `static func`/`case` requirements, I figured I might as well just rework the `RouterAction` types as an `enum` and get all the case path stuff "for free".

In all, this means that using `TCARouter` now looks like this:

```swift
TCARouter(store.scope(state: \.routes, action: \.router)) { screen in
  SwitchStore(screen) { state in
    switch state { ... }
  }
}
```

Another thing that's been changed to bring it closer towards TCA is the `forEachRoute` reducer methods.
Since there's no `RouterAction` protocol any more, I found that I had to pass in a `CaseKeyPath` for the action. Then, I realised I might as well just pass in a `WritableKeyPath` for the state as well, to match the TCA syntax for the `forEach` reducer. With that gone, there's not really any need for the `IdentifiedRouterState` and `IndexedRouterState` protocols, so I have deprecated them with a warning that they're obsolete. I might not have used the *most* correct deprecation warning on those though! That said, I think it'll flag up to any users to remove the protocol conformance for their states.

I have tested these changes as far back as iOS 13.0. Navigation doesn't entirely work on iOS 13.0, but appears to work fine on iOS 13.7. Not tried on versions in-between. Everything seems to work great on iOS 15, 16, and 17 though. I've tested my branch against a private project which makes heavy use of TCACoordinators, and that all seemed to work too, including previously written unit tests.

These changes ultimately give the library a path to supporting Observation in a subsequent release, as that relies on the use of cached `KeyPath` scoping for stores.

A fair bit of the documentation still needs updating, but I wanted to get this PR properly over to you now!

### CODE REVIEW

**Comment by johnpatrickmorgan** [comment]
Thanks so much @rhysm94! I can see how much work has gone into these changes, and it's much appreciated. It took me a while to understand everything fully, but now I've groked everything, all the changes look good to me. I'm planning to rewrite the README and add a Migration doc too. I also had a play at trying to incorporate soft deprecation of the older APIs on this branch, so I could understand the limitations, and I agree it would be more trouble than it's worth. I might check if the inverse is any easier though: i.e., introducing the new APIs on top of the old implementation, as a precursor to these changes. Thanks again!

**Comment by rhysm94** [comment]
Hi @johnpatrickmorgan

Thanks for looking this over! I did look into adding a version of the `TCARouter` initialiser that adds its own generics for `CoordinatorState` and `CoordinatorAction`, but since the initialiser would need to change anyway, and I don't think I've got a way to easily map two input `CaseKeyPath`s into the *single* `CaseKeyPath` for the `RouterAction`, I don't think it's something worth pursuing. Anyone who needs to older syntax can remain on the older version of TCACoordinators.

I'll have a look at tidying/re-adding any documentation tomorrow, but is there anything new you wanted added or anywhere in particular you can see it's now lacking?

**Comment by johnpatrickmorgan** [comment]
Thanks @rhysm94! I played around to see if there was an easy way to soft deprecate the old API (so that users have the option to upgrade one reducer at a time), but I think the simplest thing would be for users to manually add a casepath to any existing actions they're not yet ready to fully update:

```swift
// Quick update for an action that formerly conformed to `IdentifiedRouterAction`.
enum Action: CasePathable {
  case updateRoutes(IdentifiedArrayOf<Route<Screen.State>>)
  case routeAction(Screen.State.ID, action: Screen.Action)

  static var allCasePaths = AllCasePaths()

  struct AllCasePaths {
    var router: AnyCasePath<Action, IdentifiedRouterAction<Screen.State, Screen.Action>> {
      AnyCasePath { routerAction in
        switch routerAction {
        case let .routeAction(id, action):
          return .routeAction(id, action: action)
        case let .updateRoutes(newRoutes):
          return .updateRoutes(IdentifiedArray(uniqueElements: newRoutes))
        }
      } extract: { action in
        switch action {
        case let .routeAction(id, action: action):
          return .routeAction(id: id, action: action)
        case let .updateRoutes(newRoutes):
          return .updateRoutes(newRoutes.elements)
        }
      }
    }
  }
}
``` 

With that in mind, I'm planning to write a migration doc today. 

> I'll have a look at tidying/re-adding any documentation tomorrow, but is there anything new you wanted added or anywhere in particular you can see it's now lacking?

No, there are definitely some holes in the code documentation but those were already there, so they shouldn't hold up merging this PR. Thanks!

**Comment by rhysm94** [comment]
Ah! I really like that solution!

File: README.md [hunk 1]
```diff
@@ -100,21 +104,11 @@ struct Coordinator: Reducer {
       case .routeAction(_, .numberDetail(.goBackTapped)):
```
[the remark this answers is not in the document]
Comment (johnpatrickmorgan) [reply README.md]: Good spot, thanks!

**Comment by johnpatrickmorgan** [comment]
Thanks @rhysm94, I've added some docs and made the parameter names more consistent for the `forEachRoute` functions. I think this is ready to merge, unless you see any issues with my changes?

**Comment by rhysm94** [comment]
> Thanks @rhysm94, I've added some docs and made the parameter names more consistent for the `forEachRoute` functions. I think this is ready to merge, unless you see any issues with my changes?

All looks good to me! Just one extra thing that I think needs calling out:

```swift
// Before:
TCARouter(store)

// After:
TCARouter(store.scope(state: \.routes, action: \.router))
``` 

Also, just wondered about the version number. Is it fine to be a breaking change since it's pre-1.0, and migration is pretty simple anyway? Is the intention that the subsequent Observation support will be 0.10 or 1.0?

**Comment by johnpatrickmorgan** [comment]
> All looks good to me! Just one extra thing that I think needs calling out

Oh yes, thanks, I'll add that. 🙏🏼 

> Also, just wondered about the version number. Is it fine to be a breaking change since it's pre-1.0, and migration is pretty simple anyway? Is the intention that the subsequent Observation support will be 0.10 or 1.0?

Yes, I think this change can go into a 0.9 release and the Observation support can go into 0.10 release.

**Comment by johnpatrickmorgan** [comment]
Merged and shipped as v0.9.0. Thanks so much @rhysm94! 🙏

**Pull request merged**